### PR TITLE
[Metrics] Fix dead actors count is incorrect because the actor metadata is GC'ed as soon as the job is done

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

It's incorrect because as soon as the job is terminated we garbage collect GcsActor. This is supposed to happen after 5 minutes based on the existing logic. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 1]
```diff
@@ -1338,23 +1338,25 @@ void GcsActorManager::OnJobFinished(const JobID &job_id) {
       }
 
       run_delayed_(
-          [this, non_detached_actors = std::move(non_detached_actors)]() {
+          [this,
+           job_id = std::move(job_id),
+           non_detached_actors = std::move(non_detached_actors)]() {
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:1343]: It's the bug

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 2]
```diff
@@ -417,9 +417,9 @@ void GcsServer::InitGcsActorManager(const GcsInitData &gcs_init_data) {
         gcs_placement_group_manager_->CleanPlacementGroupIfNeededWhenActorDead(actor_id);
       },
       [this](std::function<void(void)> fn, boost::posix_time::milliseconds delay) {
-        boost::asio::deadline_timer timer(main_service_);
-        timer.expires_from_now(delay);
-        timer.async_wait([fn](const boost::system::error_code &error) {
+        auto timer = std::make_shared<boost::asio::deadline_timer>(main_service_);
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_server.cc:420]: I found this logic actually never worked before lol... the reason is the timer is GC'ed as soon as this block is exited which makes the async_wait aborted. So this logic never worked. cc @scv119 I will also see if there's other usage like this from the codebase as a follow up,.

cc @ericl actually should we just completely remove this GC actors when the job terminates? We already have logic to clean up dead actor metadata by the # of dead actors count. https://github.com/ray-project/ray/blob/073e7bc04d989607848552537f9f5ac91fa07d85/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L1479

**Comment by rkooo567** [comment]
Working on test failures...

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc (hunk 1 above)
Comment (rickyyx) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:1343]: how late does this one run btw?

**Review by rickyyx — APPROVED** [review]
<div><br/>via <a href="https://giphy.com/thedailyshow/">The Daily Show with Trevor Noah</a> on <a href="https://giphy.com/gifs/thedailyshow-thumbs-up-good-job-great-xT9DPPniA6EjHhmLqU">GIPHY</a></div>

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc (hunk 1 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:1343]: 5m now!

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 2 above)
Comment (ericl) [reply src/ray/gcs/gcs_server/gcs_server.cc:420]: I think I agree it makes sense to remove this logic if we already have another mechanism to remove actors, but I don't have the full context.

**Review by ericl — COMMENTED** [review]
Note that we don't necessarily need to mess with the GCS actor cleanup to fix the metric reporting. We could avoid decrementing DEAD instead, e.g.:


```
diff --git a/src/ray/gcs/gcs_server/gcs_actor_manager.h b/src/ray/gcs/gcs_server/gcs_actor_manager.h
index 405fa77a8e..4b559bd934 100644
--- a/src/ray/gcs/gcs_server/gcs_actor_manager.h
+++ b/src/ray/gcs/gcs_server/gcs_actor_manager.h
@@ -133,11 +133,13 @@ class GcsActor {
 
   ~GcsActor() {
     if (last_metric_state_) {
-      RAY_LOG(DEBUG) << "Decrementing state at "
-                     << rpc::ActorTableData::ActorState_Name(last_metric_state_.value())
-                     << " " << GetActorTableData().class_name();
-      counter_->Decrement(
-          std::make_pair(last_metric_state_.value(), GetActorTableData().class_name()));
+      if (last_metric_state_.value() != DEAD) {
+        RAY_LOG(DEBUG) << "Decrementing state at "
+                       << rpc::ActorTableData::ActorState_Name(last_metric_state_.value())
+                       << " " << GetActorTableData().class_name();
+        counter_->Decrement(
+            std::make_pair(last_metric_state_.value(), GetActorTableData().class_name()));
+      }
     }
   }
```

**Comment by rkooo567** [comment]
@ericl I think it is better doing that approach and fix this issue separately. I will modify the PR and cherry pick that change only.

**Comment by rkooo567** [comment]
Replacement https://github.com/ray-project/ray/pull/31019

**Pull request closed without merging**